// ryuu-fca-api/lib/userAgent.js
const browsers = {
    desktop: {
        windows: {
            platform: 'Windows NT 10.0; Win64; x64',
            browsers: {
                chrome: ['122.0.0.0', '121.0.0.0'],
                firefox: ['123.0', '122.0'],
                edge: ['122.0.2365.92']
            }
        }
    },
    mobile: {
        android: {
            devices: ['Samsung Galaxy S24', 'Google Pixel 8', 'Xiaomi 14'],
            versions: ['14', '13'],
            browsers: ['Chrome']
        }
    }
};

const getRandom = arr => arr[Math.floor(Math.random() * arr.length)];

const generateUA = () => {
    const isDesktop = Math.random() < 0.8;
    
    if (isDesktop) {
        const platform = browsers.desktop.windows;
        const browserName = getRandom(Object.keys(platform.browsers));
        const version = getRandom(platform.browsers[browserName]);
        
        return browserName === 'firefox' 
            ? `Mozilla/5.0 (${platform.platform}) Gecko/20100101 Firefox/${version}`
            : `Mozilla/5.0 (${platform.platform}) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/${version} Safari/537.36`;
    }
    
    const device = getRandom(browsers.mobile.android.devices);
    const version = getRandom(browsers.mobile.android.versions);
    
    return `Mozilla/5.0 (Linux; Android ${version}; ${device}) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Mobile Safari/537.36`;
};

module.exports = { generateUA };
